hal/armv7m/mpu.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix-RTOS loader
+ *
+ * MPU routines
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <lib/lib.h>
+#include <lib/log.h>
+#include <lib/errno.h>
+#include <hal/hal.h>
+#include <syspage.h>
+#include "mpu.h"
+
+
+enum { mpu_fError = 0x01, mpu_fConfigured = 0x04, mpu_fValid = 0x08 };
+
+
+struct {
+	syspage_hal_t hal;
+	u8 mapsCnt;
+	u8 regCnt;
+	u8 regMax;
+	u8 flags;
+} mpu_common;
+
+
+/* Binary count trailing zeros */
+__attribute__((always_inline)) static inline u8 _ctz(u32 val)
+{
+	asm volatile(" \
+		rbit %0, %0; \
+		clz %0, %0"
+		: "+l"(val));
+	return val;
+}
+
+
+static void mpu_attrPrint(u32 attr, int enabled)
+{
+	static const char *strAP[8] = { "na/na", "rw/na", "rw/ro", "rw/rw", "reserved", "ro/ro", "ro/ro" };
+	static const char strYesNo[2] = { 'n', 'y' };
```

Comment:
You don't have to define number of elements in array if they are initialized when declared.

Reply:
Don't want flex array here, currently the limit tells that there is no more than number of elements. So, I am not changing that.